apps/emqx/src/emqx_persistent_session_ds_state.erl

```diff
@@ -372,6 +375,19 @@ clear_will_message_now(SessionId) when is_binary(SessionId) ->
 clear_will_message(Rec) ->
     set_will_message(undefined, Rec).
 
+-spec set_offline_info_now(emqx_persistent_session_ds:id(), _Info :: map()) -> ok.
+set_offline_info_now(SessionId, Info) ->
+    transaction(fun() ->
+        case kv_restore(?session_tab, SessionId) of
+            [Metadata0] ->
+                Metadata = Metadata0#{?offline_info => Info},
+                kv_persist(?session_tab, SessionId, Metadata),
+                ok;
+            [] ->
+                ok
+        end
+    end).
```

Comment:
"transaction" is implemented via `async_dirty`:
https://github.com/emqx/emqx/pull/12980/files#diff-4f926b90b373b839679213fbdb3c30c784f5575bc3d4b721afce21466dded77eR738

It means that `read` -> `modify` -> `persist` sequence is not protected from the race condition whatsoever, and can potentially get interleave with other process (e.g. sessions itself). This is why ownership rules must not be violated.